### PR TITLE
SNOW-11: PortalMenuWidget

### PR DESCRIPTION
**Pull request by sarahelsaig**
[SNOW-11](https://lombiq.atlassian.net/browse/SNOW-11)

### CODE REVIEW

File: Extensions/Widgets/Drivers/PortalMenuWidgetDisplayDriver.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+using OrchardCore.DisplayManagement.Handlers;
+using OrchardCore.DisplayManagement.Views;
+using OrchardCore.Navigation;
+using System.Threading.Tasks;
+using static Lombiq.HelpfulExtensions.Extensions.Widgets.WidgetTypes;
+using static Lombiq.HelpfulLibraries.Libraries.Navigation.PortalNavigationProviderBase;
+
+namespace Lombiq.HelpfulExtensions.Extensions.Widgets
+{
+    public class PortalMenuWidgetDisplayDriver : ContentDisplayDriver
```
Comment (barthamark) [inline Extensions/Widgets/Drivers/PortalMenuWidgetDisplayDriver.cs]: I'll leave here my comment I wrote on the other PRs:

> I think the word Portal is too specific. With this widget, you are providing the ability to render navigation anywhere. Maybe SiteNavigationWidgetFilter would be more appropriate.

File: Views/PortalMenuWidget.cshtml [hunk 2]
```diff
@@ -0,0 +1,84 @@
+@using Lombiq.HelpfulExtensions.Extensions.Widgets
+@using OrchardCore.Navigation
+
+<style asp-name="font-awesome"></style>
+
+@{
+    var useWrapper = !(Model.NoWrapper as bool? ?? false);
+    IEnumerable<MenuItem> menuItems = Model.MenuItems;
+
+    const string blockName = WidgetTypes.PortalMenuWidget;
```
Comment (barthamark) [inline Views/PortalMenuWidget.cshtml]: The convention for constants here is PascalCased or at least this is how we did it in other cshtml files.

**Review by barthamark — CHANGES_REQUESTED** [review]

File: Views/PortalMenuWidget.cshtml (hunk 2 above)
Comment (0liver) [reply Views/PortalMenuWidget.cshtml]: The PascalCase naming of local constants has always bothered me, because it's not consistent with other local variables. I'm in favor of using CamelCase like here, but I guess that would be another bike-shed we might want to avoid right now.

Comment (barthamark) [reply Views/PortalMenuWidget.cshtml]: I'm not really bothered about all these, honestly, just pointed out that this is the first time I see these in cshtml files. I'm okay to use camelCase. What are your thoughts @DAud-IcI ?

Comment (sarahelsaig) [reply Views/PortalMenuWidget.cshtml]: It's a convention you have to actively remember to go against other more generic conventions (also Rider shouts at me) and I forgot, so I'm not a fan of it. I say let's go behind that bike-shed Oliver mentioned and duke it out ;)
On a more serious note, do we have this documented somewhere? If not, where should we?

Comment (sarahelsaig) [reply Views/PortalMenuWidget.cshtml]: Also if I turn the constant PascalCase, then it's weird with the similarly named camelCase variable next to it, for example:
`<a id="@dropdownIdName" class=" @DropdownElementName"` (sample pruned to better drive the point home)

File: Extensions/Widgets/Drivers/PortalMenuWidgetDisplayDriver.cs (hunk 1 above)
Comment (sarahelsaig) [reply Extensions/Widgets/Drivers/PortalMenuWidgetDisplayDriver.cs]: The widget's template is very front page main menu specific, with the Bootstrap hamburger menu and all. In principle it can be rendered anywhere, but I can't think of any other use-case than as a portal menu. What do you have in mind?

File: Views/PortalMenuWidget.cshtml (hunk 2 above)
Comment (0liver) [reply Views/PortalMenuWidget.cshtml]: Now, that we have a project-wide agreement, we can bring it to the General Developers table 😀

**Comment by barthamark** [comment]
@Piedone what's up with these errors in GH actions?

**Comment by Piedone** [comment]
You can ignore these for now.

**Comment by Piedone** [comment]
I fixed this to avoid confusion, please merge from dev.

File: Views/PortalMenuWidget.cshtml (hunk 2 above)
Comment (barthamark) [reply Views/PortalMenuWidget.cshtml]: Agreed.

**Review by barthamark — APPROVED** [review]

File: Extensions/Widgets/ViewModels/MenuWidgetViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        public IEnumerable<MenuItem> MenuItems { get; set; }
+
+        public MenuWidgetViewModel(bool noWrapper = false, IEnumerable<MenuItem> menuItems = null)
+        {
+            NoWrapper = noWrapper;
+            MenuItems = menuItems ?? Enumerable.Empty<MenuItem>();
+        }
+
+        public MenuWidgetViewModel(dynamic model)
+            : this(model.NoWrapper as bool? ?? false, model.MenuItems as IEnumerable<MenuItem>)
```
Comment (0liver) [inline Extensions/Widgets/ViewModels/MenuWidgetViewModel.cs:19]: @DAud-IcI What do you think of:
```suggestion
            : this((bool)(model.NoWrapper == true), model.MenuItems as IEnumerable<MenuItem>)
```

File: Lombiq.HelpfulExtensions.csproj [hunk 4]
```diff
@@ -24,12 +24,8 @@
...
 
   <ItemGroup Condition="'$(NuGetBuild)' != 'True'">
-    <ProjectReference
-      Include="..\..\Libraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries.csproj"
-      Condition="Exists('..\..\Libraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries.csproj')" />
-    <ProjectReference
-      Include="..\..\Libraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries.csproj"
-      Condition="Exists('..\..\Libraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries.csproj')" />
+    <ProjectReference Include="..\..\Libraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries.csproj" Condition="Exists('..\..\Libraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries.csproj')" />
+    <ProjectReference Include="..\..\Libraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries.csproj" Condition="Exists('..\..\Libraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries.csproj')" />
```
Comment (0liver) [inline Lombiq.HelpfulExtensions.csproj:28]: Why was this necessary?

File: Readme.md [hunk 5]
```diff
@@ -37,6 +37,7 @@ Includes:
 - ContainerWidget: Works as a container for further widgets. It has a FlowPart attached to it so it can contain additional widgets as well.
 - HtmlWidget: Adds HTML editing and displaying capabilities using a WYSIWYG editor. 
 - LiquidWidget: Adds Liquid code editing and rendering capabilities.
+- MenuWidget: Renders a Bootstrap navigation menu using the provided `MenuItem`s as a widget.
```
Comment (0liver) [inline Readme.md:40]: For clarity I would reorder the sentence like this:
```suggestion
- MenuWidget: Renders a Bootstrap navigation menu as a widget using the provided `MenuItem`s.
```

File: Readme.md [hunk 6]
```diff
@@ -37,6 +37,7 @@ Includes:
 - ContainerWidget: Works as a container for further widgets. It has a FlowPart attached to it so it can contain additional widgets as well.
 - HtmlWidget: Adds HTML editing and displaying capabilities using a WYSIWYG editor.
```
Comment (0liver) [inline Readme.md:38]:
```suggestion
- HtmlWidget: Adds HTML editing and displaying capabilities using a WYSIWYG editor.
```

File: Views/MenuWidget.cshtml [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    const string dropdownElementName = blockName + "__dropdown";
+    const string dropdownItemElementName = blockName + "__dropdownItem";
+
+    var contentIdName = FormattableString.Invariant($"{contentElementName}_{Guid.NewGuid()}");
+    var dropdownIdName = FormattableString.Invariant($"{dropdownElementName}_{Guid.NewGuid()}");
+}
+
+@if (!viewModel.NoWrapper)
+{
+    @:<nav class="navbar navbar-expand-md">
```
Comment (0liver) [inline Views/MenuWidget.cshtml:21]: @DAud-IcI Why do you need the `@:` here? The markup itself should interrupt the code block, no?

File: Views/MenuWidget.cshtml [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+                                </li>
+                            }
+                        </ul>
+                    }
+                    else
+                    {
+                        var activeClass = Context.Request.Path == menuItem.Href?.Partition("?").Left ? "active" : string.Empty;
+                        <li class="nav-item">
+                            <a id="@menuItem.Id"
+                               class="nav-link @activeClass @string.Join(" ", menuItem.Classes)"
```
Comment (0liver) [inline Views/MenuWidget.cshtml:67]: Wouldn't the following be sooo much nicer?
```suggestion
                               class="nav-link @activeClass @menuItem.Classes.Join()"
```

File: Extensions/Widgets/ViewModels/MenuWidgetViewModel.cs (hunk 3 above)
Comment (sarahelsaig) [reply Extensions/Widgets/ViewModels/MenuWidgetViewModel.cs:19]: It confuses the IDE (it thinks the `== true` is not necessary even though it is) so I'd rather not, to avoid future problems.

File: Lombiq.HelpfulExtensions.csproj (hunk 4 above)
Comment (sarahelsaig) [reply Lombiq.HelpfulExtensions.csproj:28]: No idea.

File: Views/MenuWidget.cshtml [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+                               @if (!menuItem.LocalNav)
+                               {
+                                   <text>target="_blank"</text>
+                               }
+                               href="@(menuItem.Href ?? "#")">@menuItem.Text</a>
```
Comment (0liver) [inline Views/MenuWidget.cshtml:72]: @DAud-IcI Why would we render a `#` for an empty link? Why not leave the attribute empty?

See https://www.w3.org/2001/tag/2011/01/HashInURI-20110115#NakedHash for some input.

Comment (0liver) [reply Views/MenuWidget.cshtml:72]: There are some interesting answers [here](https://stackoverflow.com/q/5637969/177710), too. To be clear, I've never been a fan of `#` in `href`s.

File: Views/MenuWidget.cshtml (hunk 7 above)
Comment (sarahelsaig) [reply Views/MenuWidget.cshtml:21]: You get an error that "Tag nav is not closed", becuase it thinks the HTML isn't well nested.

Comment (0liver) [reply Views/MenuWidget.cshtml:21]: 😲

File: Views/MenuWidget.cshtml (hunk 8 above)
Comment (sarahelsaig) [reply Views/MenuWidget.cshtml:67]: I think I wrote this prior.

File: Extensions/Widgets/ViewModels/MenuWidgetViewModel.cs (hunk 3 above)
Comment (0liver) [reply Extensions/Widgets/ViewModels/MenuWidgetViewModel.cs:19]: Now I remember - this is what I had in mind:

```suggestion
            : this(model.NoWrapper as bool? == true, model.MenuItems as IEnumerable<MenuItem>)
```

I find this more intuitive than the `?? false`. What do you think?

File: Views/MenuWidget.cshtml (hunk 9 above)
Comment (sarahelsaig) [reply Views/MenuWidget.cshtml:72]: For some reason I thought that would not render the attribute at all. Maybe that's what happens when the value is null, or is that in Vuejs? Not sure.

Anyway, It's fixed, I committed it into `issue/SNOW-12` since this PR is already merged. While I have no issues with additional reviews please limit it to the ones that are still open, otherwise it's a pain.

Comment (0liver) [reply Views/MenuWidget.cshtml:72]: But you know what? Have the `#`. We have more important things to do. It's just a fallback, anyways.

File: Extensions/Widgets/ViewModels/MenuWidgetViewModel.cs (hunk 3 above)
Comment (sarahelsaig) [reply Extensions/Widgets/ViewModels/MenuWidgetViewModel.cs:19]: Sure, that works. Though you also need to add parenthesis.

Comment (0liver) [reply Extensions/Widgets/ViewModels/MenuWidgetViewModel.cs:19]: I didn't have to in LINQPad:

image

😕

Comment (sarahelsaig) [reply Extensions/Widgets/ViewModels/MenuWidgetViewModel.cs:19]: Yeah it's a code analysis warning. You don't have those in LINQPad.

File: Views/MenuWidget.cshtml (hunk 9 above)
Comment (0liver) [reply Views/MenuWidget.cshtml:72]: Sure thing. Will stick to open issues.
